### PR TITLE
[nnyeah] Check dependencies in output

### PR DESCRIPTION
**Pull request by stephen-hawley**
This PR includes:
- change visibility of methods in `Program.cs` to make it unit-testable
- infrastructure to find the Microsoft platform dlls
- tests to ensure that Xamarin and Microsoft dlls are where we expect
- a (failing) test that shows that dependency is not being removed

### CODE REVIEW

File: tools/nnyeah/tests/utils/Compiler.cs [hunk 1]
```diff
@@ -18,6 +18,7 @@ public enum PlatformName {
 
 	public class Compiler {
 		const string MonoCompiler = "/Library/Frameworks/Mono.framework/Versions/Current/Commands/csc";
+		const string buildDirectory = "../../../../../../_build";
```
Comment (rolfbjarne) [inline tools/nnyeah/tests/utils/Compiler.cs:21]: The Configuration class already contains all the logic you're adding:

https://github.com/xamarin/xamarin-macios/blob/92eda7f35399f442f8730aa539571ad8fd44d7da/tests/common/Configuration.cs#L14

It should be possible to just add that file to your project (although you might have to add more files too).

Other comments point to more specific locations in the Configuration class.

File: tools/nnyeah/tests/utils/Compiler.cs [hunk 2]
```diff
@@ -75,13 +76,38 @@ static string PlatformLibPath (PlatformName platformName, string libName)
...
+			platformName switch {
+				PlatformName.macOS => Path.Combine (buildDirectory, "Microsoft.macOS.Runtime.osx-arm64/runtimes/osx-arm64/lib/net6.0"),
+				PlatformName.iOS => Path.Combine (buildDirectory, "Microsoft.iOS.Runtime.ios-arm/runtimes/ios-arm/lib/net6.0"),
+				PlatformName.tvOS => Path.Combine (buildDirectory, "Microsoft.tvOS.Runtime.tvossimulator-x64/runtimes/tvossimulator-x64/lib/net6.0"),
+				PlatformName.watchOS => throw new NotImplementedException (),
+				_ => throw new NotImplementedException (),
+			};
+
+		public static string MicrosoftLibName (PlatformName platformName) =>
+			platformName switch {
```
Comment (rolfbjarne) [inline tools/nnyeah/tests/utils/Compiler.cs:101]: https://github.com/xamarin/xamarin-macios/blob/92eda7f35399f442f8730aa539571ad8fd44d7da/tests/common/Configuration.cs#L707-L723

File: tools/nnyeah/tests/utils/Compiler.cs [hunk 3]
```diff
@@ -75,13 +76,38 @@ static string PlatformLibPath (PlatformName platformName, string libName)
...
 				PlatformName.iOS => "Xamarin.iOS",
 				PlatformName.tvOS => "Xamarin.TVOS",
 				PlatformName.watchOS => "Xamarin.WatchOS",
 				_ => throw new NotImplementedException (),
 			};
+
+		public static string XamarinPlatformLibraryPath (PlatformName platformName) =>
+			Path.Combine (XamarinPlatformLibDirectory (platformName), $"{XamarinLibName (platformName)}.dll");
+
+		public static string MicrosoftPlatformLibDirectory (PlatformName platformName) =>
```
Comment (rolfbjarne) [inline tools/nnyeah/tests/utils/Compiler.cs:91]: You should be able to use the reference assemblies:

https://github.com/xamarin/xamarin-macios/blob/92eda7f35399f442f8730aa539571ad8fd44d7da/tests/common/Configuration.cs#L514-L521

**Review by rolfbjarne — CHANGES_REQUESTED** [review]

File: tools/nnyeah/tests/utils/Compiler.cs [hunk 4]
```diff
@@ -75,13 +76,38 @@ static string PlatformLibPath (PlatformName platformName, string libName)
...
+			platformName switch {
+				PlatformName.macOS => Path.Combine (buildDirectory, "Microsoft.macOS.Runtime.osx-arm64/runtimes/osx-arm64/lib/net6.0"),
+				PlatformName.iOS => Path.Combine (buildDirectory, "Microsoft.iOS.Runtime.ios-arm/runtimes/ios-arm/lib/net6.0"),
+				PlatformName.tvOS => Path.Combine (buildDirectory, "Microsoft.tvOS.Runtime.tvossimulator-x64/runtimes/tvossimulator-x64/lib/net6.0"),
+				PlatformName.watchOS => throw new NotImplementedException (),
+				_ => throw new NotImplementedException (),
```
Comment (tj-devel709) [inline tools/nnyeah/tests/utils/Compiler.cs:97]: Would it be helpful to throw in the name of the platform?
```suggestion
				_ => throw new NotImplementedException (nameof (platformName)),
```

File: tools/nnyeah/tests/utils/Compiler.cs [hunk 5]
```diff
@@ -75,13 +76,38 @@ static string PlatformLibPath (PlatformName platformName, string libName)
...
+			};
+
+		public static string MicrosoftLibName (PlatformName platformName) =>
+			platformName switch {
+				PlatformName.macOS => "Microsoft.macOS",
+				PlatformName.iOS => "Microsoft.iOS",
+				PlatformName.tvOS => "Xamarin.tvOS",
+				PlatformName.watchOS => throw new NotImplementedException (),
+				_ => throw new NotImplementedException (),
```
Comment (tj-devel709) [inline tools/nnyeah/tests/utils/Compiler.cs:106]: same here?

File: tools/nnyeah/tests/utils/Compiler.cs (hunk 1 above)
Comment (rolfbjarne) [reply tools/nnyeah/tests/utils/Compiler.cs:22]: This `buildDirectory` variable doesn't seem to be needed anymore:

```suggestion
```

**Review by rolfbjarne — APPROVED** [review]

**Review by mandel-macaque — APPROVED** [review]

**Pull request merged**